pontoon/base/models.py

```diff
@@ -3260,6 +3261,40 @@ def stats(self, project, paths, locale):
 
         return translated_resources.aggregated_stats()
 
+    def update_stats(self, locales):
+        """Update stats on a list of TranslatedResource.
+        """
+        translated_resources = list(self)
```

Comment:
You should only convert `translated_resources` to a list within `bulk_update`, otherwise it will have negative effect on performance. That will also allow you to prefetch related objects needed for faster calculations.

We should remove the `locales` parameter, because you can get it from the `TranslatedResources`, same as you already do for projects. I'd change the way you retrieve them though, and also use the same logic for collecting `ProjectLocales`.

```python
def update_stats(self):
    self = self.prefetch_related('resource__project', 'locale')

    locales = Locale.objects.filter(
        translatedresources__in=self,
    ).distinct()

    projects = Project.objects.filter(
        resources__translatedresources__in=self,
    ).distinct()

    projectlocales = ProjectLocale.objects.filter(
        project__resources__translatedresources__in=self,
        locale__translatedresources__in=self,
    ).distinct()

    for translated_resource in self:
        translated_resource.calculate_stats(save=False)

    # bulk update TR

    # aggregate stats for P, L and PL
```